[MNT] Parallelize Example Notebook Runs in CI

satwiksps opened the pull request:
## Reference Issues/PRs

- Closes #2666

## What does this implement/fix? Explain your changes.

This PR updates the Continuous Integration (CI) workflows to run example notebooks in parallel, significantly reducing the execution time for the `run-notebook-examples` job.

**Specific changes:**
1.  **Updated `.github/utilities/run_examples.sh`**: 
    -   Added logic to detect the number of available CPU cores (using `nproc` on Linux and `sysctl` on macOS).
    -   Implemented parallel execution using `xargs -P` when the new `MULTITHREADED` argument is set to `true`.
    -   Refactored the script to separate notebook collection from execution logic so exclusion rules apply consistently.
2.  **Updated `.github/workflows/pr_examples.yml`**: 
    -   Passed `true` as the second argument to `run_examples.sh` to enable parallel execution on Pull Requests.
3.  **Updated `.github/workflows/periodic_tests.yml`**: 
    -   Passed `true` as the second argument to `run_examples.sh` to enable parallel execution on nightly builds.

**Results:**
The logs confirm that multithreading is successfully enabled.

1.  **Parallel Execution Triggered:** The script correctly detects available cores and executes `xargs` with the `-P` flag (e.g., `-P 3` on macOS runners).
2.  **Concurrent Processing:** The logs show multiple notebooks starting the conversion process simultaneously (interleaved `Converting notebook ...` messages) rather than waiting for the previous one to finish writing, which was the behavior in the sequential version.

**Log Snippet:**
```text
+ sysctl -n hw.ncpu
+ CORES=3
+ echo 'Running 44 notebooks in parallel on 3 cores...'
+ xargs -0 -n 1 -P 3 bash -c '$CMD "$1"' _
...
[NbConvertApp] Converting notebook examples/classification/rotation_forest.ipynb ...
[NbConvertApp] Converting notebook examples/classification/early_classification.ipynb ...
[NbConvertApp] Converting notebook examples/classification/classification.ipynb ...
```

## Does your contribution introduce a new dependency? If yes, which one?

No.

## Any other comments?

N/A

## PR checklist

### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/aeon-toolkit/aeon/blob/main/.all-contributorsrc). Alternatively, you can use the [@all-contributors](https://allcontributors.org/docs/en/bot/usage) bot to do this for you **after** the PR has been merged.
- [x] The PR title starts with either [ENH], [MNT], [DOC], [BUG], [REF], [DEP] or [GOV] indicating whether the PR topic is related to enhancement, maintenance, documentation, bugs, refactoring, deprecation or governance.

### For new estimators and functions
- [ ] I've added the estimator/function to the online [API documentation](https://www.aeon-toolkit.org/en/latest/api_reference.html).
- [ ] (OPTIONAL) I've added myself as a `__maintainer__` at the top of relevant files and want to be contacted regarding its maintenance. Unmaintained files may be removed. This is for the full file, and you should not add yourself if you are just making minor changes or do not want to help maintain its contents.

### For developers with write access
- [ ] (OPTIONAL) I've updated aeon's [CODEOWNERS](https://github.com/aeon-toolkit/aeon/blob/main/CODEOWNERS) to receive notifications about future changes to these files.

MatthewMiddlehurst reviewed (changes requested):
Seems slightly faster which is good. There are a few things though:

1. Add something in the script for Windows, even if its just doing it without multithreading with a warning
2. Temporarily break one or two of the notebooks so we know failures are still caught in the CI as they should be
3. We no longer get the runtime for each notebook when running, is there any way to keep these?

satwiksps commented:
@MatthewMiddlehurst. I have updated the script to solve this. Since parallel xargs processes run in sub-shells and cannot update main script variables, I implemented a log aggregation strategy using a temporary directory. This allows us to capture the runtime and status of every parallel worker and generate the summary report at the end.

```
Run .github/utilities/run_examples.sh true true
Skipping: examples/classification/interval_based.ipynb
Skipping: examples/classification/shapelet_based.ipynb
Skipping: examples/classification/deep_learning.ipynb
Skipping: examples/classification/convolution_based.ipynb
Skipping: examples/classification/feature_based.ipynb
Skipping: examples/classification/dictionary_based.ipynb
Skipping: examples/classification/distance_based.ipynb
Skipping: examples/classification/hybrid.ipynb
Skipping: examples/transformations/signature_method.ipynb
Skipping: examples/datasets/load_data_from_web.ipynb
Skipping: examples/clustering/partitional_clustering.ipynb
Skipping: examples/benchmarking/regression.ipynb
Skipping: examples/benchmarking/reference_results.ipynb
Skipping: examples/benchmarking/published_results.ipynb
Skipping: examples/similarity_search/code_speed.ipynb
Running 44 notebooks in parallel on 3 cores...
Finished: examples/classification/early_classification.ipynb (10s) [PASS]
Finished: examples/classification/rotation_forest.ipynb (15s) [PASS]
Finished: examples/transformations/preprocessing.ipynb (23s) [PASS]
Finished: examples/anomaly_detection/anomaly_detection.ipynb (29s) [PASS]
Finished: examples/transformations/sast.ipynb (7s) [PASS]
Finished: examples/transformations/smoothing_filters.ipynb (6s) [PASS]
Finished: examples/transformations/catch22.ipynb (15s) [PASS]
Finished: examples/transformations/rocket.ipynb (8s) [PASS]
Finished: examples/transformations/transformations.ipynb (13s) [PASS]
Finished: examples/transformations/imbalance.ipynb (5s) [PASS]
Finished: examples/transformations/tsfresh.ipynb (45s) [PASS]
Finished: examples/transformations/minirocket.ipynb (19s) [PASS]
Finished: examples/classification/classification.ipynb (100s) [PASS]
Finished: examples/segmentation/segmentation.ipynb (3s) [PASS]
Finished: examples/transformations/channel_selection.ipynb (7s) [PASS]
Finished: examples/datasets/data_loading.ipynb (4s) [PASS]
Finished: examples/datasets/provided_data.ipynb (6s) [PASS]
Finished: examples/datasets/datasets.ipynb (5s) [PASS]
Finished: examples/segmentation/hidalgo_segmentation.ipynb (25s) [PASS]
Finished: examples/datasets/data_unequal.ipynb (12s) [PASS]
Finished: examples/clustering/clustering.ipynb (2s) [PASS]
Finished: examples/segmentation/segmentation_with_clasp.ipynb (45s) [PASS]
Finished: examples/visualisation/plotting_for_learning_tasks.ipynb (5s) [PASS]
Finished: examples/clustering/feature_based_clustering.ipynb (54s) [PASS]
Finished: examples/visualisation/plotting_series.ipynb (8s) [PASS]
Finished: examples/visualisation/plotting_results.ipynb (6s) [PASS]
Finished: examples/visualisation/plotting_distances.ipynb (6s) [PASS]
Finished: examples/benchmarking/benchmarking.ipynb (3s) [PASS]
Finished: examples/networks/deep_learning.ipynb (87s) [PASS]
Finished: examples/regression/regression.ipynb (15s) [PASS]
Finished: examples/distances/distances.ipynb (9s) [PASS]
Finished: examples/forecasting/iterative.ipynb (9s) [PASS]
Finished: examples/distances/sklearn_distances.ipynb (19s) [PASS]
Finished: examples/forecasting/setar.ipynb (6s) [PASS]
Finished: examples/forecasting/forecasting.ipynb (4s) [PASS]
Finished: examples/forecasting/direct.ipynb (5s) [PASS]
Finished: examples/forecasting/theta.ipynb (6s) [PASS]
Finished: examples/forecasting/ets.ipynb (3s) [PASS]
Finished: examples/visualisation/plotting_estimators.ipynb (117s) [PASS]
Finished: examples/forecasting/regression.ipynb (32s) [PASS]
Finished: examples/similarity_search/distance_profiles.ipynb (3s) [PASS]
Finished: examples/base/base_classes.ipynb (4s) [PASS]
Finished: examples/similarity_search/similarity_search.ipynb (15s) [PASS]
Finished: examples/base/series_estimator.ipynb (8s) [PASS]
TIME	STATUS	NOTEBOOK
117s	PASS	examples/visualisation/plotting_estimators.ipynb
100s	PASS	examples/classification/classification.ipynb
87s	PASS	examples/networks/deep_learning.ipynb
54s	PASS	examples/clustering/feature_based_clustering.ipynb
45s	PASS	examples/transformations/tsfresh.ipynb
45s	PASS	examples/segmentation/segmentation_with_clasp.ipynb
32s	PASS	examples/forecasting/regression.ipynb
29s	PASS	examples/anomaly_detection/anomaly_detection.ipynb
25s	PASS	examples/segmentation/hidalgo_segmentation.ipynb
23s	PASS	examples/transformations/preprocessing.ipynb
19s	PASS	examples/transformations/minirocket.ipynb
19s	PASS	examples/distances/sklearn_distances.ipynb
15s	PASS	examples/transformations/catch22.ipynb
15s	PASS	examples/similarity_search/similarity_search.ipynb
15s	PASS	examples/regression/regression.ipynb
15s	PASS	examples/classification/rotation_forest.ipynb
13s	PASS	examples/transformations/transformations.ipynb
12s	PASS	examples/datasets/data_unequal.ipynb
10s	PASS	examples/classification/early_classification.ipynb
9s	PASS	examples/forecasting/iterative.ipynb
9s	PASS	examples/distances/distances.ipynb
8s	PASS	examples/visualisation/plotting_series.ipynb
8s	PASS	examples/transformations/rocket.ipynb
8s	PASS	examples/base/series_estimator.ipynb
7s	PASS	examples/transformations/sast.ipynb
7s	PASS	examples/transformations/channel_selection.ipynb
6s	PASS	examples/visualisation/plotting_results.ipynb
6s	PASS	examples/visualisation/plotting_distances.ipynb
6s	PASS	examples/transformations/smoothing_filters.ipynb
6s	PASS	examples/forecasting/theta.ipynb
6s	PASS	examples/forecasting/setar.ipynb
6s	PASS	examples/datasets/provided_data.ipynb
5s	PASS	examples/visualisation/plotting_for_learning_tasks.ipynb
5s	PASS	examples/transformations/imbalance.ipynb
5s	PASS	examples/forecasting/direct.ipynb
5s	PASS	examples/datasets/datasets.ipynb
4s	PASS	examples/forecasting/forecasting.ipynb
4s	PASS	examples/datasets/data_loading.ipynb
4s	PASS	examples/base/base_classes.ipynb
3s	PASS	examples/similarity_search/distance_profiles.ipynb
3s	PASS	examples/segmentation/segmentation.ipynb
3s	PASS	examples/forecasting/ets.ipynb
3s	PASS	examples/benchmarking/benchmarking.ipynb
2s	PASS	examples/clustering/clustering.ipynb

```

MatthewMiddlehurst reviewed (approved):
thanks
